File: MessagingMeshLib.NET/Logger.cs
using System;
using System.Collections.Generic;
using System.Threading;

namespace MessagingMeshLib.NET
{
    /// <summary>
    /// Not so much a logger as a log-broker.
    /// 
    /// Log messages can be logged to the Logger and clients can register
    /// for callbacks when messages are logged. The client can then show
    /// or log the messages.
    /// </summary>
    public class Logger
    {
        #region Public types

        /// <summary>
        /// Log level enum.
        /// </summary>
        public enum LogLevel
        {
            DEBUG,
            INFO,
            WARN,
            ERROR,
            FATAL
        };

        /// <summary>
        /// Function signature for log callbacks.
        /// </summary>
        public delegate void Callback(LogLevel logLevel, string message);

        #endregion

        #region Public methods

        /// <summary>
        /// Registers a function to be called when messages are logged.
        /// </summary>
        public static void registerCallback(Callback callback)
        {
            lock (m_callbacksLocker)
            {
                m_callbacks.Add(callback);
            }
        }

        /// <summary>
        /// Logs a message at the level specified.
        /// </summary>
        public static void log(LogLevel logLevel, string message)
        {
            // We take a copy of the callbacks (to avoid locking over the callbacks themselves)...
            List<Callback> callbacks;
            lock (m_callbacksLocker)
            {
                callbacks = new List<Callback>(m_callbacks);
            }

            // We convert a null message to text...
            message ??= "(null)";

            // We add info the the message...
            var threadName = Thread.CurrentThread.Name;
            var messageToLog = $"MessagingMesh ({threadName}): {message}";

            // We notify the registered callbacks...
            foreach (var callback in callbacks)
            {
                try
                {
                    callback(logLevel, messageToLog);
                }
                catch (Exception)
                {
                    // We do not do anything here. We do not log, as this is the logger!
                }
            }
        }

        /// <summary>
        /// Logs an exception including the stack trace.
        /// </summary>
        public static void log(Exception ex)
        {
            log(LogLevel.ERROR, ex.ToString());
        }

        /// <summary>
        /// Logs a message at DEBUG level.
        /// </summary>
        public static void debug(string message)
        {
            log(LogLevel.DEBUG, message);
        }

        /// <summary>
        /// Logs a message at INFO level.
        /// </summary>
        public static void info(string message)
        {
            log(LogLevel.INFO, message);
        }

        /// <summary>
        /// Logs a message at WARN level.
        /// </summary>
        public static void warn(string message)
        {
            log(LogLevel.WARN, message);
        }

        /// <summary>
        /// Logs a message at ERROR level.
        /// </summary>
        public static void error(string message)
        {
            log(LogLevel.ERROR, message);
        }

        /// <summary>
        /// Logs a message at FATAL level.
        /// </summary>
        public static void fatal(string message)
        {
            log(LogLevel.FATAL, message);
        }

        #endregion

        #region Private data

        // List of registered callbacks and an object to lock it...
        private static List<Callback> m_callbacks = new();
        private static object m_callbacksLocker = new();

        #endregion
    }
}

File: MessagingMeshLib/Logger.cpp
#include "Logger.h"
#include <memory>
#include <format>
#include "uv.h"
#include "Utils.h"
#include "UVUtils.h"
using namespace MessagingMesh;

// Static fields...
std::atomic<std::shared_ptr<const Logger::VecCallback>> Logger::m_callbacks;
std::mutex Logger::m_mutex;

// Log levels as strings...
const std::string Logger::DEBUG_STRING = "DEBUG";
const std::string Logger::INFO_STRING = "INFO";
const std::string Logger::WARN_STRING = "WARN";
const std::string Logger::ERROR_STRING = "ERROR";
const std::string Logger::FATAL_STRING = "FATAL";
const std::string Logger::UNKNOWN_LOG_LEVEL_STRING = "[UNKNOWN-LOG-LEVEL]";


// Registers a function to be called when messages are logged.
void Logger::registerCallback(Callback callback)
{
    std::lock_guard<std::mutex> lock(m_mutex);

    // Copy-on-write pattern...
    auto oldCallbacks = m_callbacks.load();
    auto newCallbacks = std::make_shared<VecCallback>(oldCallbacks ? *oldCallbacks : VecCallback{});
    newCallbacks->push_back(callback);
    m_callbacks.store(newCallbacks);
}

// Logs a message at the level specified.
void Logger::log(LogLevel logLevel, const std::string& message)
{
    // Note that we do not use the mutex here. Instead we atomically load the callbacks shared pointer
    // giving us a reference to the immutable collection of callbacks currently in effect.
    auto pCallbacks = m_callbacks.load();
    if (!pCallbacks || pCallbacks->empty())
    {
        return;
    }

    // We add info to the message...
    auto threadName = UVUtils::getThreadName();
    auto messageToLog = std::format("MessagingMesh ({}): {}", threadName, message);

    // We notify the registered callbacks...
    for (auto& callback : *pCallbacks)
    {
        try
        {
            callback(logLevel, messageToLog);
        }
        catch (...)
        {
            // We catch exceptions so that we continue logging to all callbacks even
            // if one of them throws. We do not do anything here. In particular we do 
            // not log the error, as we are the logger!
        }
    }
}

// Logs a message at DEBUG level.
void Logger::debug(const std::string& message)
{
    log(LOG_LEVEL_DEBUG, message);
}

// Logs a message at INFO level.
void Logger::info(const std::string& message)
{
    log(LOG_LEVEL_INFO, message);
}

// Logs a message at WARN level.
void Logger::warn(const std::string& message)
{
    log(LOG_LEVEL_WARN, message);
}

// Logs a message at ERROR level.
void Logger::error(const std::string& message)
{
    log(LOG_LEVEL_ERROR, message);
}

// Logs a message at FATAL level.
void Logger::fatal(const std::string& message)
{
    log(LOG_LEVEL_FATAL, message);
}

// Returns the log-level as a string.
const std::string& Logger::toString(LogLevel logLevel)
{
    switch (logLevel)
    {
    case LOG_LEVEL_DEBUG:
        return DEBUG_STRING;
    case LOG_LEVEL_INFO:
        return INFO_STRING;
    case LOG_LEVEL_WARN:
        return WARN_STRING;
    case LOG_LEVEL_ERROR:
        return ERROR_STRING;
    case LOG_LEVEL_FATAL:
        return FATAL_STRING;
    default:
        return UNKNOWN_LOG_LEVEL_STRING;
    }
}
